### PR TITLE
Updated Compact Configuration

### PR DESCRIPTION
**Pull request by yuce**
This PR:
* Changes `compact_serializers` configuration to be a list of `CompactSerializer`.
* Adds `get_class` method to the base `CompactSerializer`.
* Updates related docs and tests.
* Trivially updates `run_tests.py` to pass given comand line arguments to `pytest`.

### CODE REVIEW

File: hazelcast/config.py [hunk 1]
```diff
@@ -1369,19 +1369,15 @@ def use_public_ip(self, value):
...
-    def compact_serializers(self) -> typing.Dict[typing.Type, CompactSerializer]:
+    def compact_serializers(self) -> typing.List[CompactSerializer]:
         return self._compact_serializers
 
     @compact_serializers.setter
-    def compact_serializers(self, value: typing.Dict[typing.Type, CompactSerializer]) -> None:
-        if isinstance(value, dict):
-            for clazz, serializer in value.items():
-                if not isinstance(clazz, type):
-                    raise TypeError("Keys of compact_serializers must be classes")
```
Comment (srknzl) [inline hazelcast/config.py:1380]: We are losing the validation of classes here. Can we somehow check that getClass method exists in the serializer and it returns a class and nothing else. Does instanceof check takes care of the existence of methods ? I am sure it does not  handle the validation of the return values of getClass and getTypeName :). 

I know the `get_type_name` validation was nonexistent before as well but I think we can add it. In Node.js, because of fail fast goal of ours, I am calling the methods and checking the return value's type. See https://github.com/hazelcast/hazelcast-nodejs-client/blob/801696baf6492475ad1ddf664356a9707a0bf092/src/config/ConfigBuilder.ts#L512 and https://github.com/hazelcast/hazelcast-nodejs-client/blob/801696baf6492475ad1ddf664356a9707a0bf092/src/config/ConfigBuilder.ts#L517

File: docs/serialization.rst [hunk 2]
```diff
@@ -146,7 +146,7 @@ Then, a serializer for it can be implemented as below:
 
 .. code:: python
 
-    from hazelcast.serialization.api import CompactSerializer, CompactWriter, CompactReader
+    from hazelcast.serialization.api import CompactSerializer, CompactWriter, CompactReader, CompactSerializableClass
```
Comment (srknzl) [inline docs/serialization.rst:149]: Don't forget to change this file if you change `get_class` method's return type.

File: examples/serialization/compact_serialization_example.py [hunk 3]
```diff
@@ -1,5 +1,10 @@
 from hazelcast import HazelcastClient
-from hazelcast.serialization.api import CompactSerializer, CompactWriter, CompactReader
+from hazelcast.serialization.api import (
+    CompactSerializer,
+    CompactWriter,
+    CompactReader,
+    CompactSerializableClass,
```
Comment (srknzl) [inline examples/serialization/compact_serialization_example.py:6]: Don't forget to change this file if you change `get_class` method's return type.

File: examples/sql/sql_compact_example.py [hunk 4]
```diff
@@ -1,7 +1,12 @@
 import dataclasses
 
 from hazelcast import HazelcastClient
-from hazelcast.serialization.api import CompactSerializer, CompactWriter, CompactReader
+from hazelcast.serialization.api import (
```
Comment (srknzl) [inline examples/sql/sql_compact_example.py:4]: Don't forget to change this file if you change `get_class` method's return type.

File: hazelcast/serialization/service.py [hunk 5]
```diff
@@ -344,7 +344,7 @@ def __init__(
         self._registration_lock = threading.RLock()
         self._int_type_id = _int_type_to_type_id.get(config.default_int_type, None)
 
-        self._compact_types = set(config.compact_serializers.keys())
+        self._compact_types = [c.get_class() for c in config.compact_serializers]
```
Comment (srknzl) [inline hazelcast/serialization/service.py:347]: Does not using a dict make more sense here? We will do a lot of lookups in this data type.

File: run_tests.py [hunk 6]
```diff
@@ -21,29 +21,31 @@ def wait_until_rc_is_ready():
...
-
-                process = subprocess.run(args)
-                rc_process.kill()
-                rc_process.wait()
-                sys.exit(process.returncode)
-            except:
-                rc_process.kill()
-                rc_process.wait()
-                raise
+with open("rc_stdout.log", "w") as rc_stdout:
```
Comment (srknzl) [inline run_tests.py:24]: I think you did that becuase there is no files that import run_tests.py. Are we sure there won't be in the future? It is safe to include a guard `if __name__ == "__main__":`

File: run_tests.py [hunk 7]
```diff
@@ -21,29 +21,31 @@ def wait_until_rc_is_ready():
...
+        try:
+            wait_until_rc_is_ready()
+            sys.stdout.flush()
+
+            args = [
+                "pytest",
+                "--cov=hazelcast",
+                "--cov-report=xml",
+            ]
+            args.extend(sys.argv[1:])
```
Comment (srknzl) [inline run_tests.py:36]: Why did you add this line?

**Review by srknzl — CHANGES_REQUESTED** [review]

File: run_tests.py (hunk 6 above)
Comment (yuce) [reply run_tests.py:24]: I don't think we'll use this file as a library.

File: run_tests.py (hunk 7 above)
Comment (yuce) [reply run_tests.py:36]: To make arguments passed to `run_tests` passed to pytest, like:
```
./run_tests.py -k FilterMyTest
```

File: hazelcast/serialization/service.py (hunk 5 above)
Comment (yuce) [reply hazelcast/serialization/service.py:347]: It should be a set instead of a list. Will update.

Comment (srknzl) [reply hazelcast/serialization/service.py:347]: It is even better (O(1)) to make it a dict than a set (O(log(n)))

Comment (yuce) [reply hazelcast/serialization/service.py:347]: Are you sure set lookup is O(log(n))? According to https://wiki.python.org/moin/TimeComplexity it is O(1).

Comment (srknzl) [reply hazelcast/serialization/service.py:347]: Ok I was thinking it is like a set in cpp.

File: hazelcast/serialization/api.py [hunk 8]
```diff
@@ -3059,6 +3059,14 @@ def write(self, writer: CompactWriter, obj: CompactSerializableClass) -> None:
                 to write.
         """
 
+    @abc.abstractmethod
+    def get_class(self) -> CompactSerializableClass:
+        """Returns the class for this serializer.
+
+        Returns:
+            The class for this serializer.
```
[the remark this answers is not in the document]
Comment (yuce) [reply hazelcast/serialization/api.py]: Other docs for this class doesn't have examples, so I've left that out. I've made the method's purpose clearer at: https://github.com/hazelcast/hazelcast-python-client/pull/546/commits/ae2532d913aad0f7b02f20cacd0852209ea024c4

File: hazelcast/config.py (hunk 1 above)
Comment (yuce) [reply hazelcast/config.py:1380]: Since the user has to derive their class from `CompactSerializer)`, the class will have its `get_class` method.

File: docs/serialization.rst (hunk 2 above)
Comment (yuce) [reply docs/serialization.rst:149]: Removed the import for CompactSerializableClass at: https://github.com/hazelcast/hazelcast-python-client/pull/546/commits/ae2532d913aad0f7b02f20cacd0852209ea024c4

File: examples/serialization/compact_serialization_example.py (hunk 3 above)
Comment (yuce) [reply examples/serialization/compact_serialization_example.py]: Removed the import for that at: https://github.com/hazelcast/hazelcast-python-client/pull/546/commits/6f6d8fdd8681b6349a84526961a55ef96e1a4eed

File: examples/sql/sql_compact_example.py (hunk 4 above)
Comment (yuce) [reply examples/sql/sql_compact_example.py:4]: At: https://github.com/hazelcast/hazelcast-python-client/pull/546/commits/6f6d8fdd8681b6349a84526961a55ef96e1a4eed

File: docs/serialization.rst (hunk 2 above)
Comment (srknzl) [reply docs/serialization.rst:149]: Looks like you have not removed the import yet.

```suggestion
    from hazelcast.serialization.api import CompactSerializer, CompactWriter, CompactReader
```

File: hazelcast/config.py (hunk 1 above)
Comment (srknzl) [reply hazelcast/config.py:1380]: What about return value validation?

Comment (yuce) [reply hazelcast/config.py:1380]: That's something to be caught by the type checker.

File: docs/serialization.rst (hunk 2 above)
Comment (yuce) [reply docs/serialization.rst]: My comments keep disappearing. Removed that at: https://github.com/hazelcast/hazelcast-python-client/pull/546/commits/df32f6dae22693344c393081f1a7339517051218

File: hazelcast/config.py (hunk 1 above)
Comment (srknzl) [reply hazelcast/config.py:1380]: However not everyone uses a type checker in python. 

Think of Node.js client. For non-ts users we do a runtime check even we don't have to. Isn't this what we want to do, failing fast? 

In java, we the compilation step validates the type. 

If we have the goal of failing fast given a  wrong configuration, we should do validate here.

Comment (srknzl) [reply hazelcast/config.py:1380]: We will discuss later how much to validate. Skipping now.

**Review by srknzl — APPROVED** [review]

File: docs/serialization.rst [hunk 9]
```diff
@@ -149,26 +149,29 @@ Then, a serializer for it can be implemented as below:
...
+
 The last step is to register the serializer in the client configuration.
 
 .. code:: python
 
     client = HazelcastClient(
-        compact_serializers={
-            Employee: EmployeeSerializer(),
-        }
+        compact_serializers=[
```
Comment (utku-caglayan) [inline docs/serialization.rst:172]: Can you update the documentation for this parameter? It looks like user should pass a dictionary, based on that.

**Review by utku-caglayan — DISMISSED** [review]

File: docs/serialization.rst (hunk 9 above)
Comment (yuce) [reply docs/serialization.rst:172]: No, the user passes a list of serializers.

**Review by srknzl — APPROVED** [review]

**Review by utku-caglayan — APPROVED** [review]

File: docs/serialization.rst (hunk 9 above)
Comment (yuce) [reply docs/serialization.rst:172]: Thanks for catching the outdated documentation. Fixed at: https://github.com/hazelcast/hazelcast-python-client/pull/546/commits/878d5ed0a636f79ab81034483582c1767837806d

**Review by srknzl — APPROVED** [review]

**Review by utku-caglayan — APPROVED** [review]
